syntax = "proto3";

import "google/protobuf/empty.proto";

package minitwitter;

service MiniTwitter {
  rpc SendMessage(Message) returns (google.protobuf.Empty);
  rpc GetMessages(GetMessagesRequest) returns (GetMessagesResponse);
  rpc GetAttachment(GetAttachmentsRequest) returns (GetAttachmentsResponse);
  rpc AddLike(AddLikeRequest) returns (google.protobuf.Empty);
  rpc AddComment(AddCommentRequest) returns (google.protobuf.Empty);
}

message Message {
  string text = 1;
  string sender = 2;
  string creation_time = 3;
  optional FileAttachment file_attachment = 4;
}

message FileAttachment {
  string file_name = 1;
  bytes file_data = 2;
  string file_type = 3;
  string file_data_id = 4;
}

message GetMessagesRequest {
  int32 n = 1;
}

message GetMessagesResponse {
  repeated Message messages = 1;
}

message GetAttachmentsRequest {
  string attachment_id = 1;
}

message GetAttachmentsResponse {
  FileAttachment attachments = 1;
}

message AddLikeRequest {
  string message_id = 1;
  string username = 2;
}

message AddCommentRequest {
  string message_id = 1;
  string username = 2;
  string text = 3;
}
